### PR TITLE
FHIRPath CLI #1283

### PR DESCRIPTION
**Pull request by prb112**
Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: docs/src/pages/guides/FHIRValidationGuide.md [hunk 1]
```diff
@@ -224,3 +224,124 @@ The IBM FHIR Server provides a basic implementation of the $validate operation t
...
+```
+
+Without the `--pretty` option, the results are: 
+
+```
+1
+1000001
+```
+
+For instace, to run the fhir-path-cli using a string.
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRValidationGuide.md:268]:
```suggestion
For instance, to run the fhir-path-cli using a string.
```

File: docs/src/pages/guides/FHIRValidationGuide.md [hunk 2]
```diff
@@ -224,3 +224,124 @@ The IBM FHIR Server provides a basic implementation of the $validate operation t
...
+
+``` shell
+java -jar fhir-path-4.6.0-cli.jar --path 'text.status.value' --type string --resource '{
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRValidationGuide.md:271]: I wonder if we could get it to work over specific fhir data types.  they don't have an official serialization, but we do support `toString` on them in our model.  just food for thought on the `--resource` flag name.

File: docs/src/pages/guides/FHIRValidationGuide.md [hunk 3]
```diff
@@ -224,3 +224,124 @@ The IBM FHIR Server provides a basic implementation of the $validate operation t
...
 Coded elements (code, Coding, CodeableConcept data types), maybe have a binding element that specifies a ValueSet that that element is bound to. This means that the coded element must have a value that comes from that value set. The FHIR profile component will expand value sets according to the ValueSet [expansion algorithm](http://hl7.org/fhir/valueset.html#expansion) for ValueSets that include CodeSystem resources available via the FHIR registry component.
+
+# fhir-path-cli - command line execution of fhirpath over a FHIR Resource
+
+The fhir-path-cli enables commandline processing of a FHIR Path over a FHIR resource and outputs the results. 
+
+The options for the command line utility are:
+- `--path 'fhir-path'`
+- `--format ['json'|'xml']` Default is 'json'
+- `--type ['stdin'|'file'|'string']`
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRValidationGuide.md:235]: can `type` have a default too?  maybe `stdin` or `file`...whichever we think will be most common

Comment (lmsurpre) [reply docs/src/pages/guides/FHIRValidationGuide.md:235]: actually, maybe we can just remove this and deduce it from the other arguments?
* `stdin` can be the default
* if `--file` is present then its type=file
* if `--resource` is present then its type=string

Comment (prb112) [reply docs/src/pages/guides/FHIRValidationGuide.md:235]: I thought about that too. I decided to adhere to what was in the issue. 
Maybe we infer if it doesn't exist?

File: docs/src/pages/guides/FHIRValidationGuide.md (hunk 2 above)
Comment (prb112) [reply docs/src/pages/guides/FHIRValidationGuide.md:271]: interesting

File: fhir-path/src/main/java/com/ibm/fhir/path/tools/Main.java [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+import com.ibm.fhir.path.visitor.FHIRPathNodeVisitor;
+
+/**
+ * Processes the resource from one of the entry points, into an in memory resource
+ * a FHIR Path is executed over it, and the output is put out to the screen.
+ *
+ * The types of entry points are:
+ * a file
+ * a string (raw command line arg); or
+ * stdin (so we can pipe to it)
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/tools/Main.java:53]: javadoc will squash the whitespace here; i recommend putting it in a list
```suggestion
 * <ul>
 * <li>a file
 * <li>a string (raw command line arg); or
 * <li>stdin (so we can pipe to it)
 * </ul>
```

File: docs/src/pages/guides/FHIRValidationGuide.md (hunk 3 above)
Comment (lmsurpre) [reply docs/src/pages/guides/FHIRValidationGuide.md:235]: I fine with that.  Although I can't imagine anyone wanting to specify it when they don't have to!

Comment (prb112) [reply docs/src/pages/guides/FHIRValidationGuide.md]: Fixed... interpolating the type from the command parameters.

File: docs/src/pages/guides/FHIRValidationGuide.md (hunk 2 above)
Comment (prb112) [reply docs/src/pages/guides/FHIRValidationGuide.md]: maybe a separate issue? we could park it in a value in a Parameters object and call it.

**Comment by prb112** [comment]
Rebuild depends on PR that fixes the snapshot

File: docs/src/pages/guides/FHIRValidationGuide.md (hunk 2 above)
Comment (lmsurpre) [reply docs/src/pages/guides/FHIRValidationGuide.md]: I'd say don't even worry about this one right now.  I just know there were times where I wanted to execute FHIRPath against some specific backbone element and, while thats supported by our FHIREvaluator, it won't be supported from the cli with this design.

Comment (lmsurpre) [reply docs/src/pages/guides/FHIRValidationGuide.md]: For it to work the way I was sort of hoping for, we'd need to be able to parse a specific Element / Element sub-type (independent of its resource wrapper).  But I'm pretty sure we don't support that and its probably not even worth entertaining that with a separate issue at this point.

File: fhir-path/src/main/java/com/ibm/fhir/path/tools/Main.java [hunk 5]
```diff
@@ -0,0 +1,372 @@
+/*
+ * (C) Copyright IBM Corp. 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.fhir.path.tools;
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/tools/Main.java:7]: in other projects, I think I used `*.tool` in the package (singular vs plural).  Not sure how important it is to be consistent on this though

**Review by lmsurpre — APPROVED** [review]
LGTM

**Comment by lmsurpre** [comment]
Anything we need to do from the release automation side to ensure that both the cli.jar and the normal jar get uploaded?

File: fhir-path/src/main/java/com/ibm/fhir/path/tools/Main.java (hunk 5 above)
Comment (prb112) [reply fhir-path/src/main/java/com/ibm/fhir/path/tools/Main.java:7]: Interesting... I looked in another project and it had tools.  might be a good cleanup item

**Pull request merged**